### PR TITLE
Use a non-contiguous Dictionary for tests

### PR DESCRIPTION
**Pull request by benaadams**
To cover test hole found in https://github.com/dotnet/coreclr/pull/17196#discussion_r177871393

### CODE REVIEW

**Comment by benaadams** [comment]
Fails heavily locally for me in `.CopyTo`, so lets see how this goes

**Comment by benaadams** [comment]
The tests are a bit intertwined

**Comment by danmosemsft** [comment]
@dotnet-bot test this please (kusto was down)

**Comment by benaadams** [comment]
Unrelated failures

```
OSX.1012.Amd64.Open-x64-Debug

System.Net.Tests.HttpListenerRequestTests
ContentLength_GetProperty_ReturnsExpected(method: \"PUT\", contentLengthString: \"Content-Length: 1\", expected: 1, hasEntityBody: True)

Unhandled Exception of Type System.Net.Internals.SocketExceptionFactory+ExtendedSocketException
Message :
System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Connection refused 127.0.0.1:1330
Stack Trace :
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 4421
   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 832
   at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 947
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 958
   at System.Net.Sockets.Socket.Connect(String host, Int32 port) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 902
   at System.Net.Tests.HttpListenerFactory.GetConnectedSocket() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.HttpListener/tests/HttpListenerFactory.cs:line 155
   at System.Net.Tests.HttpListenerRequestTests..ctor() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Net.HttpListener/tests/HttpListenerRequestTests.cs:line 23
   at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, Boolean wrapExceptions) in /Users/buildagent/agent/_work/482/s/src/mscorlib/src/System/RtType.cs:line 4833
```
```
OSX.1013.Amd64.Open-x64-Debug

System.ComponentModel.Composition.DirectoryCatalogTests
LoadedFiles_NonStaticallyReferencedAssembly

Unhandled Exception of Type Xunit.Sdk.NotEmptyException
Message :
Assert.NotEmpty() Failure
Stack Trace :
   at System.ComponentModel.Composition.DirectoryCatalogTests.LoadedFiles_NonStaticallyReferencedAssembly() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/Hosting/DirectoryCatalogTests.cs:line 356
```

```
Windows.10.Amd64.Open-x86-Release

ProxyAuth_Digest_Succeeds
System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest

Unhandled Exception of Type System.TimeoutException
Message :
System.TimeoutException : WhenAllOrAnyFailed timed out after 60000
Stack Trace :
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in D:\j\workspace\windows-TGrou---f8ac6754\src\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs:line 53
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest.<>c__DisplayClass114_0.<<ProxyAuth_Digest_Succeeds>b__0>d.MoveNext() in D:\j\workspace\windows-TGrou---f8ac6754\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs:line 3014
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in D:\j\workspace\windows-TGrou---f8ac6754\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 67
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest.ProxyAuth_Digest_Succeeds() in D:\j\workspace\windows-TGrou---f8ac6754\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs:line 3000
--- End of stack trace from previous location where exception was thrown ---
```

**Comment by benaadams** [comment]
test OSX x64 Debug Build
test Windows x86 Release Build

**Comment by benaadams** [comment]
Will do a followup PR for the Enumerator+Remove test https://github.com/dotnet/coreclr/pull/17196#issuecomment-377360621

File: src/System.Collections/tests/Generic/Dictionary/Dictionary.Tests.cs [hunk 1]
```diff
@@ -164,6 +164,67 @@ public void CopyConstructorExceptions()
...
+        public void CopyTo_NonContiguousDictionary(int count)
+        {
+            Dictionary<string, string> collection = (Dictionary<string, string>)CreateDictionary(count, k => k.ToString());
+            string[] array = new string[count];
+            collection.Keys.CopyTo(array, 0);
+            int i = 0;
+            foreach (KeyValuePair<string, string> obj in collection)
+                Assert.Equal(array[i++], obj.Key);
+
+            collection.Values.CopyTo(array, 0);
```
Comment (danmosemsft) [inline src/System.Collections/tests/Generic/Dictionary/Dictionary.Tests.cs]: Ideally would be separate test for Keys and Values?

**Pull request merged**